src/proxy/metadata.rs

```diff
@@ -0,0 +1,182 @@
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use anyhow::anyhow;
+use bytes::Bytes;
+use http_body_util::Full;
+use hyper::service::service_fn;
+use hyper::{Response, StatusCode};
+use serde::Serialize;
+use tokio::net::TcpStream;
+use tracing::{error, trace};
+
+use traffic::ConnectionOpen;
+
+use crate::identity::Identity;
+use crate::metrics::traffic::ConnectionClose;
+use crate::metrics::{traffic, MetricGuard, Metrics};
+
+use super::Error;
+
+#[derive(Clone, Copy, Hash, Debug, Eq, PartialEq)]
+pub struct ConnectionTuple {
+    pub src: SocketAddr,
+    pub dst: SocketAddr,
+}
+
+/// ConnectionTracker trackers connections. It stores a map of all active connections, and handles recording
+/// metrics for connection establishment and termination.
+#[derive(Clone)]
+pub struct ConnectionTracker {
+    conns: Arc<Mutex<HashMap<ConnectionTuple, Option<Identity>>>>,
+    metrics: Arc<Metrics>,
+}
+
+impl ConnectionTracker {
+    pub fn new(metrics: Arc<Metrics>) -> Self {
+        ConnectionTracker {
+            conns: Default::default(),
+            metrics,
+        }
+    }
+
+    /// fetch looks up a tuple and returns the connection metadata
+    pub fn fetch(&self, ctu: &ConnectionTuple) -> Option<ConnectionMetadata> {
+        let cm = self.conns.lock().unwrap();
+        cm.get(ctu)
+            .cloned()
+            .map(|identity| ConnectionMetadata { identity })
+    }
+
+    pub fn track<'a>(
+        &'a self,
+        ct: ConnectionTuple,
+        conn: &'a ConnectionOpen,
+    ) -> ConnectionGuard<'a> {
+        {
+            let mut c = self.conns.lock().unwrap();
+
+            let id = if let Some(ds) = &conn.derived_source {
+                ds.identity.clone()
+            } else {
+                conn.destination.as_ref().map(|w| w.identity())
+            };
+            trace!(tuple=?ct, id=?id, "tracking connection");
+            c.insert(ct, id);
+        }
+        let connection_close = self.metrics.increment_defer::<_, ConnectionClose>(conn);
+        ConnectionGuard::<'a> {
+            _metric: connection_close,
+            ct: self,
+            tup: ct,
+        }
+    }
+}
+
+pub struct ConnectionGuard<'a> {
+    /// _metric is just here to drop when we drop and record the connection close metric
+    _metric: MetricGuard<'a, ConnectionClose<'a>>,
+    ct: &'a ConnectionTracker,
+    tup: ConnectionTuple,
+}
+
+impl<'a> Drop for ConnectionGuard<'a> {
+    fn drop(&mut self) {
+        trace!(tuple=?self.tup, "forgetting connection");
+        let mut cm = self.ct.conns.lock().unwrap();
+        cm.remove(&self.tup);
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct ConnectionMetadata {
+    #[serde(default)]
+    identity: Option<Identity>,
+}
+
+/// METADATA_SERVER_IP provides the well-known metadata server IP.
+/// This is captured by the redirection.
+pub const METADATA_SERVER_IP: IpAddr = IpAddr::V4(Ipv4Addr::new(169, 254, 169, 111));
```

Comment:
Can we just hardcode a specific metadata server IP in OSS ztunnel like this?

If it's "well-known", to whom is it well-known?

Reply:
I mean we _can_ - the IP is irrelevant other than the client and server need to be in agreement about it and it should not conflict with other IPs users want to use for other purposes.

> and would prefer a metadata port (like prometheus, xds, etc etc) so we don't need to rely on known IP keying in zt proper.

What is better about a special port instead of a special IP? Or did you want special IP + special port (which seems good?)

Reply:
I guess I'm not clear on why ztunnel needs a special IP at all?

If we can redirect any pod outbound on <whatever IP you want> to <whatever ztunnel you want> and <whatever ztunnel port we want> outside of ztunnel itself - why does ztunnel need to match a well-known IP?

Reply:
Is it any simpler to have it in the CNI layer instead of in ztunnel though?

In general there is a tension of what goes at which layer. For example, we could skip. ztunnel entirely in cases where we know it won't do anything but pass through.

So far we have leaned towards keeping the CNI simpler so there is a better chance we can replace it with better approaches (and rust is easier than iptables/eBPF)

Reply:
`Is it any simpler to have it in the CNI layer instead of in ztunnel though?` 

Long term I think so, since the CNI (ours or someone else's) will be handling/redirecting the traffic based on intended dest anyway, and we already have the concept of directing packets to different well-known ztunnel listener ports and handling them differently as a result (for purposes of early exit/handling bypass in eBPF/iptables redirect logic, for instance)

`So far we have leaned towards keeping the CNI simpler so there is a better chance we can replace it with better approaches (and rust is easier than iptables/eBPF)`

Absolutely a fan of keeping CNI simpler, but sending captured traffic based on intended dest to a given ztunnel listener is what CNI is for - that's what it does best. It's how we do DNS redir for instance. CNI redir already does special-casing around certain dest ports (or whether src is host ip, etc etc) for purposes of data path optimization - it's already handling this stuff.

Feels logically consistent to maintain that pattern here, and it has the added benefit of making it very easy for the IP to be changed or redefined without ztunnel caring (or us needing to add a flag to ztunnel to let people define what they want).

Reply:
I strongly disagree - routing traffic based on dst is what  ztunnel does.

CNI is responsible to redirect pod egress to ztunnel - it does not know where to send ( unless ztunnel and WDS/PTR-DS is integrated into the CNI ).  Knowing where to redirect an IP and how is based on Istiod configs.

Reply:
And if we want at some point to have the 'redirect all egress from pod to daemon set' - shared by all
mesh implementations and all CNI providers - we need to keep the requirements on the CNI layer as
small as possible. 

We may add extra requirements - like exclude ports for proxyless - but getting into 'route IP address x in a different way' is too much.

Reply:
> I strongly disagree - routing traffic based on dst is what ztunnel does.

> CNI is responsible to redirect pod egress to ztunnel - it does not know where to send ( unless ztunnel and WDS/PTR-DS is integrated into the CNI ). Knowing where to redirect an IP and how is based on Istiod configs.

CNI knows what ztunnel listener to redirect traffic to, and ztunnel knows what dest to send traffic to - correct - using CNI to redirect MDS traffic to ztunnel listener would be the former, which is definitely what CNI is responsible for doing today.


> And if we want at some point to have the 'redirect all egress from pod to daemon set' - shared by all
mesh implementations and all CNI providers - we need to keep the requirements on the CNI layer as
small as possible.

Valid. I'm not sure it's a realistic or practical goal, given that no two CNIs are really the same and we want to keep ztunnel from being a kitchen-sink, but we can live in hope.

> We may add extra requirements - like exclude ports for proxyless - but getting into 'route IP address x in a different way' is too much.

It doesn't strike me as significantly different than `route IP port X in a different way` which CNI already does and will always do, but I'm not set on it.